modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C

```diff
@@ -22,497 +22,682 @@ ADLAROMANCEStressUpdateBase::validParams()
-    const std::vector<std::vector<std::vector<ADReal>>> & polynomial_inputs,
-    const std::vector<std::vector<std::vector<ADReal>>> & dpolynomial_inputs,
-    std::vector<ADReal> & rom_outputs,
-    std::vector<ADReal> & drom_outputs)
+    const std::vector<ADReal> & precomputed,
+    const std::vector<std::vector<ADReal>> & polynomial_inputs,
+    const std::vector<ADReal> & dpolynomial_inputs,
+    const bool derivative)
 {
-  for (unsigned int i = 0; i < _num_outputs; ++i)
+  ADReal rom_output = 0.0;
+  ADReal val;
+  for (unsigned int c = 0; c < _num_coefs; ++c)
   {
-    rom_outputs[i] = 0.0;
-    drom_outputs[i] = 0.0;
-    for (unsigned int j = 0; j < _num_coefs; ++j)
-    {
-      ADReal xvals = coefs[i][j];
-      ADReal dxvals = 0.0;
-      for (unsigned int k = 0; k < _num_inputs; ++k)
-      {
-        xvals *= polynomial_inputs[i][k][_makeframe_helper[j][k]];
-        dxvals += dpolynomial_inputs[i][k][_makeframe_helper[j][k]];
-      }
-
-      rom_outputs[i] += xvals;
-      drom_outputs[i] += dxvals * xvals;
-    }
+    if (!derivative)
+      rom_output += precomputed[c] *
+                    polynomial_inputs[_stress_input_index]
+                                     [_makeframe_helper[c + _num_coefs * _stress_input_index]];
+    else
+      rom_output += precomputed[c] *
+                    dpolynomial_inputs[_makeframe_helper[c + _num_coefs * _stress_input_index]];
   }
+  return rom_output;
 }
 
-void
-ADLAROMANCEStressUpdateBase::convertOutput(const Real dt,
-                                           const std::vector<ADReal> & old_input_values,
-                                           const std::vector<ADReal> & rom_outputs,
-                                           const std::vector<ADReal> & drom_outputs,
-                                           std::vector<ADReal> & input_value_increments,
-                                           std::vector<ADReal> & dinput_value_increments)
+ADReal
+ADLAROMANCEStressUpdateBase::convertOutput(const std::vector<Real> & old_input_values,
+                                           const ADReal & rom_output,
+                                           const unsigned out_index,
+                                           const ADReal & drom_output,
+                                           const bool derivative)
 {
-  for (unsigned int i = 0; i < _num_outputs; ++i)
+  if (out_index == _strain_output_index)
   {
-    input_value_increments[i] = std::exp(rom_outputs[i]) * dt;
-    if (i == 0 || i == 1)
-    {
-      dinput_value_increments[i] = 0.0;
-      input_value_increments[i] *= -old_input_values[i];
-    }
+    if (derivative)
+      return std::exp(rom_output) * _dt * drom_output;
     else
-      dinput_value_increments[i] = input_value_increments[i] * drom_outputs[i];
+      return std::exp(rom_output) * _dt;
   }
+
+  if (derivative)
+    return 0.0;
+
+  ADReal expout = std::exp(rom_output);
+  if (expout > 1.0e-10)
+    expout -= 1.000000E-10;
```

Comment:
We should at least explain what the magic numbers are in a future revision of this.

Reply:
These numbers are used by Aaron in his ROM development...somewhat magical, but more tied into how those coefficients are created.